### PR TITLE
Use ReadOnlySpan and ReadOnlyMemory for reducing memory copy

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by erikzhang** [comment]
Is someone reviewing it?

File: src/neo-vm/Types/Boolean.cs [hunk 1]
```diff
@@ -31,7 +31,7 @@ public override bool Equals(StackItem other)
             {
                 return false;
             }
-            return Unsafe.MemoryEquals(GetByteArray(), bytes_other);
+            return Unsafe.MemoryEquals(GetByteArray().Span, bytes_other.Span);
```
Comment (devhawk) [inline src/neo-vm/Types/Boolean.cs]: you shouldn't need an unsafe method to iterate a span

**Comment by devhawk** [comment]
these seem like good changes, but I don't think they will reduce memory copies very much. Changing byte[] to ReadOnlyMemory<byte> doesn't affect when copies are made, just the view into the copy. For example, Integer.GetByteArray still makes a byte array copy since it's calling BigInteger.ToByteArray, even though it's now returning a ReadOnlyMemory<byte> struct.

**Comment by erikzhang** [comment]
The main improvements are for PUSHBYTES and PUSHDATA, as well as SUBSTR.

File: src/neo-vm/Types/Boolean.cs (hunk 1 above)
Comment (erikzhang) [reply src/neo-vm/Types/Boolean.cs]: This unsafe implementation is faster.

**Comment by shargon** [comment]
> Is someone reviewing it?

I come back from my vacations in 5 days

**Comment by erikzhang** [comment]
@neo-project/core I have to merge it after 24 hours if no one is reviewing it.

File: src/neo-vm/ExecutionEngine.cs [hunk 2]
```diff
@@ -217,7 +215,7 @@ private bool ExecuteInstruction()
                     // Push value
                     case OpCode.PUSH0:
                         {
-                            context.EvaluationStack.Push(EmptyBytes);
+                            context.EvaluationStack.Push(ReadOnlyMemory<byte>.Empty);
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs]: This var could be static?

File: src/neo-vm/ExecutionEngine.cs [hunk 3]
```diff
@@ -50,8 +50,6 @@ public class ExecutionEngine : IDisposable
 
         #endregion
 
-        private static readonly byte[] EmptyBytes = new byte[0];
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs]:
```
private static readonly ReadOnlyMemory<byte> EmptyBytes = ReadOnlyMemory<byte>.Empty;
```

Comment (shargon) [reply src/neo-vm/ExecutionEngine.cs]: Or maybe 
```
private static readonly StackItem EmptyBytes = new ByteArray(ReadOnlyMemory<byte>.Empty);
```

**Comment by shargon** [comment]
The coverage report doesn't appear?

File: src/neo-vm/ExecutionEngine.cs (hunk 2 above)
Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs]: `ReadOnlyMemory<byte>.Empty` is `static`.

File: src/neo-vm/ExecutionEngine.cs [hunk 4]
```diff
@@ -466,51 +465,33 @@ private bool ExecuteInstruction()
...
-                                buffer = x;
-                            }
-                            else
-                            {
-                                buffer = new byte[count];
-                                Unsafe.MemoryCopy(x, x.Length - count, buffer, 0, count);
-                            }
-                            context.EvaluationStack.Push(buffer);
+                            if (count < x.Length)
+                                x = x[^count..^0];
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs]: What it means `^`?

Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs]: `^n` == `Length - n`

File: src/neo-vm/Unsafe.cs [hunk 5]
```diff
@@ -6,12 +6,12 @@ namespace Neo.VM
     unsafe internal static class Unsafe
     {
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static bool MemoryEquals(byte[] x, byte[] y)
+        public static bool MemoryEquals(ReadOnlySpan<byte> x, ReadOnlySpan<byte> y)
         {
-            if (ReferenceEquals(x, y)) return true;
-            if (x is null || y is null) return false;
+            if (x == y) return true;
```
Comment (shargon) [inline src/neo-vm/Unsafe.cs]: Acording to the source code of `ReadOnlySpan` it doesn't check the null values
https://github.com/dotnet/coreclr/blob/340d69c20d57a2b161aa74c516886bf45df8ee22/src/System.Private.CoreLib/shared/System/ReadOnlySpan.cs#L315

So i recommend to maintain
```if (x is null || y is null) return false;```

Comment (erikzhang) [reply src/neo-vm/Unsafe.cs]: `ReadOnlySpan` is not a nullable type. It's a `struct`.

**Pull request merged**